### ISSUE TITLE
Luajit 32-bit crashing

### ISSUE THREAD

**Issue by Jehan** [msys2/MINGW-packages#9818]
Luajit 32-bit is crashing. Some logs are available here: https://p.0au.de/8c82889f/

@Wormnest was noticing that it was about an uncaught exception so we were wondering if it could be related to the grep/gcc issues a few days ago, i.e. if luajit should not just be rebuilt?

**Comment by revelator** [comment]
seems likely, try again with the rebuilt grep / gcc packages, if it works with those it would indeed seem so.

**Comment by Jehan** [comment]
Actually nowadays even the 64-bit luajit is crashing for us. Since the luajit binding is a dev feature on development version of GIMP, it's not in our priority right now and nobody really looked into this in details. So we don't know if it's something we do wrong on our side of if the problem is with the MSYS2 package.

**Comment by Wormnest** [comment]
For reference, issue in GIMP is tracked here [https://gitlab.gnome.org/GNOME/gimp/-/issues/7387](https://gitlab.gnome.org/GNOME/gimp/-/issues/7387)

**Comment by Wormnest** [comment]
```
LuaJIT 2.1.0-beta3 -- Copyright (C) 2005-2017 Mike Pall. http://luajit.org/
```

Since we use it to call plug-ins, which won't work without GIMP running, I fear it will be difficult to reproduce without GIMP.

**Comment by Jehan** [comment]
@Wormnest I think @Biswa96 mean if we can make up a simpler example reproducing the crash. But since we aren't sure yet why it crashes exactly, we can't.

**Comment by Biswa96** [comment]
No problem. I just want to narrow down the environment a little. So, how can I reproduce the issue with GIMP? Just by running it? Do I have to compile GIMP from git repository?

**Comment by Wormnest** [comment]
The problem should show up in the development version 2.99.10 available [here](https://www.gimp.org/downloads/devel/).

Run GIMP and then first open any random image.

Then go to menu Filters, Development, Goat Exercises, Exercise a Lua goat

**Comment by Biswa96** [comment]
Process Explorer shows that GIMP runs this command:

```
"C:\Users\x\AppData\Local\Programs\GIMP 2.99\bin\luajit.exe" "C:\Users\x\AppData\Local\Programs\GIMP 2.99\lib\gimp\2.99\extensions\org.gimp.extension.goat-exercises\goat-exercise-lua.lua" -gimp 270 9 7 -run 1
```

If I run that command in a separate command prompt window this error is show.

```
LibGimp:ERROR:../../libgimp/../libgimp/gimpplugin.c:267:gimp_plug_in_constructed: assertion failed: (plug_in->priv->read_channel != NULL)
Bail out! LibGimp:ERROR:../../libgimp/../libgimp/gimpplugin.c:267:gimp_plug_in_constructed: assertion failed: (plug_in->priv->read_channel != NULL)
```

Is that expected?

**Comment by Jehan** [comment]
> Is that expected?

Surely yes. It fails on reading the obviously non-existing communication channel when you try to run the plug-in standalone. GIMP plug-ins are not meant to run standalone, their whole point is to communicate with GIMP. :-)

**Comment by Wormnest** [comment]
Actually the 64-bit crash proved to be on our end, due to incorrect documentation of the closure annotations. We assume that the 32 bit crash that was first reported is also gone, so this can probably be closed.

**Comment by Jehan** [comment]
Indeed, let's close! Thanks all, sorry for the false positive. ;-)

**Comment by Biswa96** [comment]
Out of curiosity, can you provide the fixed program and information about how it was solved?

**Comment by Jehan** [comment]
It was not a package or build bug, purely a software logic bug. The problem was about introspection annotations (which tells the interpreter how to call the C library from Lua, and in particular of course how to manage memory in scripts with garbage collection).

Anyway the fix was: https://gitlab.gnome.org/GNOME/gimp/-/commit/d3e38ea4b9d1a8ddc5d92591390a64a3a940f782

And the discussion about the issue: https://gitlab.gnome.org/GNOME/gimp/-/issues/7387

Basically, we discovered that GObject Introspection documentation was wrong on closure annotations, and it was asking to do the opposite of what was to be done.

**Comment by Biswa96** [comment]
Now comments decide how to call a function 🤦‍♂️

**Comment by Jehan** [comment]
Yes. It's how GObject Introspection works. Theses annotations (a.k.a. "comments") tell the binding what is what, otherwise it can't know how to handle memory (typically in C, we handle memory manually so what we do is telling the developer what they have to do with every parameter; but in Lua, the binding needs to know when to keep an object alive, when it's safe to free, etc.).

P.S.: I thought I closed this report the other day! Apparently I hadn't! 🤷
